File: R/eval.basis.R
eval.basis <- function(evalarg, basisobj, Lfdobj=0) {
#  Computes the basis matrix evaluated at arguments in EVALARG associated
#    with basis.fd object BASISOBJ.  The basis matrix contains the values
#    at argument value vector EVALARG of applying the nonhomogeneous
#    linear differential operator LFD to the basis functions.  By default
#    LFD is 0, and the basis functions are simply evaluated at argument
#    values in EVALARG.
#
#  If LFD is a functional data object with m + 1 functions c_1, ... c_{m+1}, then it
#    is assumed to define the order m HOMOGENEOUS linear differential operator
#  Lx(t) = c_1(t) + c_2(t)x(t) + c_3(t)Dx(t) + ... + c_{m+1}D^{m-1}x(t) + D^m x(t).
#
#  If the basis type is either polygonal or constant, LFD is ignored.
#
#  Arguments:
#  EVALARG ... Either a vector of values at which all functions are to evaluated,
#              or a matrix of values, with number of columns corresponding to
#              number of functions in argument FD.  If the number of evaluation
#              values varies from curve to curve, pad out unwanted positions in
#              each column with NA.  The number of rows is equal to the maximum
#              of number of evaluation points.
#  BASISOBJ ... A basis object
#  LFDOBJ   ... A linear differential operator object
#               applied to the basis functions before they are to be evaluated.

#
#  Note that the first two arguments may be interchanged.
#
#  Last modified 26 October 2005

#  Exchange the first two arguments if the first is an BASIS.FD object
#    and the second numeric

if (is.numeric(basisobj) && inherits(evalarg, "basisfd")) {
    temp     <- basisobj
    basisobj <- evalarg
    evalarg  <- temp
}

#  check EVALARG

if (!(is.numeric(evalarg))) stop("Argument EVALARG is not numeric.")

if (!is.vector(evalarg))    stop("Argument EVALARG is not a vector.")
	
#  check basisobj
	
if (!(inherits(basisobj, "basisfd")))
  stop("Second argument is not a basis object.")

#  check LFDOBJ

Lfdobj <- int2Lfd(Lfdobj)

#  determine the highest order of derivative NDERIV required

nderiv <- Lfdobj$nderiv

#  get weight coefficient functions

bwtlist <- Lfdobj$bwtlist

#  get highest order of basis matrix

evalarray <- getbasismatrix(evalarg, basisobj, nderiv)
nbasis    <- dim(evalarray)[2]
oneb      <- matrix(1,1,nbasis)

#  Compute the weighted combination of derivatives is
#  evaluated here if the operator is not defined by an
#  integer and the order of derivative is positive.


if (nderiv > 0) {
	nonintwrd <- FALSE
	for (j in 1:nderiv) {
		bfd    <- bwtlist[[j]]
		bbasis <- bfd$basis
		if (bbasis$type != "constant" || bfd$coefs != 0) nonintwrd <- TRUE
	}
	if (nonintwrd) {
        for (j in 1:nderiv) {
            bfd   <- bwtlist[[j]]
            if (!all(c(bfd$coefs) == 0.0)) {
                wjarray   <- eval.fd(evalarg, bfd)
                Dbasismat <- getbasismatrix(evalarg, basisobj, j-1)
                evalarray <- evalarray + (wjarray %*% oneb)*Dbasismat
            }
        }
    }
}

return(evalarray)

}

File: inst/Matlab/fdaM/eval_basis.m
function [evalarray, basisobj] = eval_basis(evalarg, basisobj, Lfdobj)
%  EVAL_BASIS evaluates a basis at argument values EVALARG.
%
%  LFDOBJ is a functional data object defining the order m 
%  HOMOGENEOUS linear differential operator of the form
%  Lx(t) = w_0(t) x(t) + ... + w_{m-1}(t) D^{m-1}x(t) + D^m x(t) + 
%
%  Arguments:
%  EVALARG ... A vector of values at which all functions are to 
%              evaluated.
%  BASISOBJ ... A basis object
%  LFDOBJ   ... A linear differential operator object
%              applied to the functions that are evaluated.
%
%  Note that the first two arguments may be interchanged.
%
%  Returns:  An array of function values corresponding to the evaluation
%              arguments in EVALARG

%  Last modified 9 June 2010

if nargin < 2
    error('Number of arguments is less than 2.');
end

%  set default LFDOBJ to 0

if nargin < 3     
    Lfdobj = int2Lfd(0); 
end

%  check LFDOBJ

Lfdobj = int2Lfd(Lfdobj);

%  Exchange the first two arguments if the first is a BASIS
%    and the second numeric

if isnumeric(basisobj) && isa_basis(evalarg)
    temp     = basisobj;
    basisobj = evalarg;
    evalarg  = temp;
end

%  check EVALARG

sizeevalarg = size(evalarg);
type = getbasistype(basisobj);
if ~strcmp(type, 'FEM')
    if sizeevalarg(1) > 1 && sizeevalarg(2) > 1
        error('Argument EVALARG is not a vector.');
    end
    evalarg = evalarg(:);
else
    if size(evalarg,2) ~= 2
        error('Argument EVALARG is not a matrix with two columns.');
    end
end

%  check BASISOBJ

if ~isa_basis(basisobj)
    error ('Argument BASISOBJ is not a basis object.');
end

%  determine the highest order of derivative NDERIV required

nderiv = getnderiv(Lfdobj);

%  get weight coefficient functions

wfdcell = getwfdcell(Lfdobj);
    
%  get highest order of basis matrix

if nargout == 1
    evalarray = getbasismatrix(evalarg, basisobj, nderiv);
else
    [evalarray, basisobj] = getbasismatrix(evalarg, basisobj, nderiv);
end
nbasis    = size(evalarray,2);
onerow    = ones(1,nbasis);

%  Compute the weighted combination of derivatives is 
%  evaluated here if the operator is not defined by an 
%  integer and the order of derivative is positive.
%  Only the homogeneous part of the operator, defined by
%  cell object WFDCELL is used.

if nderiv > 0 && ~isinteger(Lfdobj)
    %  In this version, only a scalar operator is allowed.
    if size(wfdcell,1) ~= 1 && size(wfdcell,2) ~= 1
        error('WFDCELL has more than one row.');
    end
    for j = 1:nderiv
        wfdParj = wfdcell{j};
        wfdj    = getfd(wfdParj);
        wcoef   = getcoef(wfdj);
        if ~all(all(wcoef == 0.0))
            wjarray   = eval_fd(evalarg, wfdj);
            evalarray = evalarray + (wjarray*onerow).* ...
                getbasismatrix(evalarg, basisobj, j-1);
        end
    end
end


